libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -96,37 +176,128 @@ m_nameDispenser(_context.dispenser)
+		reverse(subFunctionCall.arguments.begin(), subFunctionCall.arguments.end());
+		_functionCall.arguments.emplace_back(move(subFunctionCall));
+	}
+	if (!moveInfo.returnVariableSlots.empty())
+		m_slotsForCurrentReturns = &moveInfo.returnVariableSlots;
+}
+
+void StackToMemoryMover::operator()(FunctionDefinition& _functionDefinition)
+{
+	// It is important to first visit the function body, so that it doesn't replace the memory inits for
+	// variable arguments we might generate below.
 	ASTModifier::operator()(_functionDefinition);
+
+	if (!m_functionMoveInfo.count(_functionDefinition.name))
+		return;
+
+	FunctionMoveInfo const& moveInfo = m_functionMoveInfo.at(_functionDefinition.name);
+
+	for (FunctionDefinition const* helperFunction: moveInfo.parameterHelperFunctions)
+	{
+		yulAssert(helperFunction && helperFunction->parameters.size() >= 1, "");
+		for (size_t i = 0; i < helperFunction->parameters.size() - 1; ++i)
+		{
+			yulAssert(!_functionDefinition.parameters.empty(), "");
+			_functionDefinition.parameters.pop_back();
+		}
+	}
+
+	vector<Statement> memoryVariableInits;
+
+	// All remaining function arguments that may be inaccessible in the function body, are moved
+	// at the beginning of the function body.
+	for (TypedName const& param: _functionDefinition.parameters)
+		if (auto slot = m_memoryOffsetTracker(param.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				param.location,
+				*slot,
+				Identifier{param.location, param.name}
+			);
+
+	// All memory return variables have to be initialized to zero in memory.
+	for (TypedName const& returnVariable: _functionDefinition.returnVariables)
+		if (auto slot = m_memoryOffsetTracker(returnVariable.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				returnVariable.location,
+				*slot,
+				Literal{returnVariable.location, LiteralKind::Number, "0"_yulstring, {}}
+			);
+
+	if (!memoryVariableInits.empty())
+	{
+		memoryVariableInits += move(_functionDefinition.body.statements);
+		_functionDefinition.body.statements = move(memoryVariableInits);
+	}
+
+	for (optional<YulString> slot: moveInfo.returnVariableSlots)
+		if (slot)
+			_functionDefinition.returnVariables.pop_back();
+	for (TypedName const& returnVariable: _functionDefinition.returnVariables)
+		if (auto slot = m_memoryOffsetTracker(returnVariable.name))
+			_functionDefinition.body.statements.emplace_back(
+				Assignment{
+					returnVariable.location,
+					{ Identifier{
+						returnVariable.location,
+						returnVariable.name
+					} },
+					make_unique<Expression>(
+						generateMemoryLoad(
+							m_context.dialect,
+							returnVariable.location,
+							*slot
+						))
+				}
+			);
 }
 
 void StackToMemoryMover::operator()(Block& _block)
 {
 	using OptionalStatements = std::optional<vector<Statement>>;
 	auto rewriteAssignmentOrVariableDeclaration = [&](
 		auto& _stmt,
-		auto const& _variables
+		auto& _variables
 	) -> OptionalStatements {
```

Comment:
This might be a bit simpler, if it was split into two steps, first only rewriting based on the RHS being a function call, then rewriting again based on the LHS containing variables to be moved.

Reply:
But it's a bit tricky, since the other visit's can't be run twice (especially the "assign from function argument to memory at beginning of body" thing)

Reply:
Done now in another commit, but this will need careful review for correctness.